### ISSUE TITLE
Should be filled with 0 otherwise you can get old data...

### ISSUE THREAD

**Issue by folkvir** [crypto-browserify/buffer-xor#12]
https://github.com/crypto-browserify/buffer-xor/blob/9ccb25fcb6afa96d23e96da837e7d3379475a954/index.js#L5

As noticed [in Node.js documentation](https://nodejs.org/api/buffer.html#buffer_class_method_buffer_allocunsafe_size) using allocUnsafe(length) will return a buffer of size length with possible already filled values. 
```javascript
const buf = Buffer.allocUnsafe(10);
console.log(buf);
// Prints (contents may vary): <Buffer a0 8b 28 3f 01 00 00 00 50 32>
```

Then if you XOR a Buffer full of zeroes with another buffer full of zeroes... you can potientally have undesired data...

> Note that the Buffer module pre-allocates an internal Buffer instance of size Buffer.poolSize that is used as a pool for the fast allocation of new Buffer instances created using Buffer.allocUnsafe() and the deprecated new Buffer(size) constructor only when size is less than or equal to Buffer.poolSize >> 1 (floor of Buffer.poolSize divided by two).
Use of this pre-allocated internal memory pool is a key difference between calling Buffer.alloc(size, fill) vs. Buffer.allocUnsafe(size).fill(fill). Specifically, Buffer.alloc(size, fill) will never use the internal Buffer pool, while Buffer.allocUnsafe(size).fill(fill) will use the internal Buffer pool if size is less than or equal to half Buffer.poolSize. The difference is subtle but can be important when an application requires the additional performance that Buffer.allocUnsafe() provides.

Then you should modified it to: `var buffer = Buffer.allocUnsafe(length).fill(0)`

**Comment by chiro-hiro** [comment]
@folkvir You could check this line:
https://github.com/crypto-browserify/buffer-xor/blob/9ccb25fcb6afa96d23e96da837e7d3379475a954/index.js#L8

That's mean It will be overwrite later, in this case `Buffer.allocUnsafe(length)` given better performance.

**Comment by folkvir** [comment]
@chiro-hiro You're right. It will be overwritten. But when I wrote this issue I thought the xor operator was returning an error when doing `undefined ^ Number`. Which is not the case.. Thus, I also wrote this issue according to #11. If length of a or b are not the same then, xoring 2 buffer like `xor(xor(a, b),b)` does not always return `a`. According to this, if both buffers are safe then values can be xored safely. But if the biggest (say b) has not been initialized correctly it means that all values between a.length  and b.length will be all values of b and then potentially unsafe values and also not correct values according the _xor rules_.
I don't know if there are somewhere input requirements when using xor-buffer package. 
But it has to be taking into account when using this package, especially if we don't want to check our-self all errors when xoring buffers...
I'll close this as it lacks of an answer from crypto-browserify devs. Feel free to answer me if you or the team have questions.
